src/mechanics/constraintsPde/Term.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+#include "base/Exception.h"
+#include "mechanics/nodes/NodeSimple.h"
+
+namespace NuTo
+{
+namespace ConstraintPde
+{
+//! @brief stores constraint equation terms
+class Term
+{
+public:
+    //! @brief ctor that sets all members so that the term consists of
+    //! ... + coefficient * node->GetDof(component) + ...
+    //! @param node node reference
+    //! @param component component in the dof vector of the node, has to be smaller than node.GetNumValues()
+    //! @param coefficient coefficient of the equation term
+    Term(const NodeSimple& node, int component, double coefficient)
+        : mNode(node)
+        , mComponent(component)
+        , mCoefficient(coefficient)
+    {
+        if (component >= node.GetNumValues())
+            throw Exception(__PRETTY_FUNCTION__, "Term construction failed. Node has " +
+                                                         std::to_string(node.GetNumValues()) +
+                                                         " components and you tried to constrain component " +
+                                                         std::to_string(component) + ".");
+    }
+
+    //! @brief getter for mNode
+    const NodeSimple& GetNode() const
+    {
+        return mNode;
+    }
+
+    //! @brief getter for mComponent
+    int GetComponent() const
+    {
+        return mComponent;
+    }
+
+    //! @brief getter for mCoefficient
+    double GetCoefficient() const
+    {
+        return mCoefficient;
+    }
+
+    int GetConstrainedDofNumber() const
+    {
+        return mNode.get().GetDofNumber(mComponent);
+    }
+
+private:
+    //! @brief node reference, std::reference_wrapper is neat since it provides copy
+    //! ctor and assignment
+    std::reference_wrapper<const NodeSimple> mNode;
```

Comment:
What is the advantage over a normal reference?

Reply:
Try to replace the like with a const reference and look at the compiler output. 

I tried to explain it in the comments, better believe it :) Think about a class with a const reference member. You now have to define custom copy/move ctor/assigment. And now try to write assignment operators. For me, this is not trivial and the fact that I have to provide them at all is annoying. `reference_wrapper` takes care of this.

Reply:
Hm. maybe I just misread the comment then, but still don't understand :) Why does the `reference_wrapper` provide a copy constructor that the `NodeSimple` doesn't provide? Or put differently, why is this a problem here and not at the other places where we use reference members? (Also, I've tried replacing it with a normal reference, and at least the unit test `ConstraintsPde` compiles just fine. Can't try the rest because old laptop.)

Reply:
Since I think this is more about c++ than NuTo, I chose not to change the comment in the code. The theory is [here](http://en.cppreference.com/w/cpp/language/copy_assignment). And [this](http://en.cppreference.com/w/cpp/utility/functional/reference_wrapper) is also worth reading.

The compiler (normally) implicitly creates copy/move ctor/assign for you. A const reference member (e.g. `const NodeSimple& mNode`) implicitly deletes the assignment operators for the class. (Maybe this was the missing piece of information?) This makes `Term` not assignable. Note that the default copy/move _constructor_ are implicitly created. 
If we want to use it with the STL, we can now have a look at the specific __type requirements__. E.g. [`std::vector<>::push_back`](http://en.cppreference.com/w/cpp/container/vector/push_back) that might be the only method used in `ConstraintsPde` requires the type to be `CopyInsertable` (for non language lawyers like me: = copy constructable). This is the case for `Term`, no problem here.
In other cases, e.g. `ConstraintCompanionPde` we use [`std::insert`](http://en.cppreference.com/w/cpp/container/vector/insert). This requires Copy/MoveAssignable` and will fail.

If we use a pointer `NodeSimple* mNode`, assignment is not a problem. And I think this is what `std::reference_wrapper` does, with value semantics.

> Or put differently, why is this a problem here and not at the other places where we use reference members?

1) We want `MeshFem` to be 
    - MoveConstructable (`return tmpMesh`) and 
    - MoveAssignable (`mesh = Transform(mesh, f)`). 

     Thus all classes below have to fulfill this requirement and you will see `reference_wrapper` there. The code line above might be the only place where I specifically commented on that in the code.
2) We rarely deal with values types. We mostly work on polymorphic types that have no assignment at all and require a `clone()`. Here, we do not have that problem. Either an owning `ptr_container<Interface>` or a non-owning `container<Interface*>` deals with assignments.

---
I hope this clears things up. Maybe you can add the required comments to the code that would have made it understandable without this PR comments.

Reply:
Thanks, that was very helpful indeed. What do you think of

    //! @brief node reference
    //! @remark `std::reference_wrapper` is used instead of a reference to
    //! make `Term` default CopyConstructible and CopyAssignable

which says less about what `reference_wrapper` does and maybe a bit more about what it means for the current class? I'll add a commit if you agree.

Reply:
Way better code comment, go ahead! For correctness - since a reference is `CopyConstructable` by default, we should remove this word from the comment, right? 

And while you're at it, can you fix the exception message with the _accordance_ thing as well? xD